tests/api_workflow/test_api_workflow_datasets.py

```diff
@@ -118,72 +129,206 @@ def test_get_all_datasets(self):
         assert "dataset_1" in dataset_names
         assert "shared_dataset_1" in dataset_names
 
-    def test_create_dataset_with_counter(self):
-        self.api_workflow_client.create_dataset(dataset_name="basename")
-        n_tries = 3
-        for i in range(n_tries):
-            self.api_workflow_client.create_new_dataset_with_unique_name(
-                dataset_basename="basename"
-            )
-        assert (
-            self.api_workflow_client._datasets_api.datasets[-1].name
-            == f"basename_{n_tries}"
-        )
 
-    def test_create_dataset_with_counter_nonexisting(self):
-        self.api_workflow_client.create_dataset(dataset_name="basename")
-        self.api_workflow_client.create_new_dataset_with_unique_name(
-            dataset_basename="baseName"
-        )
-        assert self.api_workflow_client._datasets_api.datasets[-1].name == "baseName"
+def test_create_new_dataset_with_unique_name__new_name(mocker: MockerFixture) -> None:
+    mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
+    mocker.patch.object(ApiWorkflowClient, "dataset_name_exists", return_value=False)
```

Comment:
> Also it avoids issues with accidentally mocking methods that do not exist on the client.
pytest raises an error when an undefined method is patched if I recall correctly.

I didn't create one helper function for the API client because it requires customisation in different test cases and thus a helper function doesn't seem to save much effort.

Reply:
Let's leave it like this for now. The benefit of the helper would be that the client is properly initialized and actually has all the instance attributes it normally has. Plus you wouldn't have to patch the full api classes but only their respective methods. So instead of this:
```
mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
client = ApiWorkflowClient()
client._datasets_api = mocker.MagicMock()
```
You could patch only the requested method:
```
mocker.patch.object(ApiWorkflowClient, "__init__", return_value=None)
client = ApiWorkflowClient()
client._datasets_api.method = mocker.MagicMock()
```

Because with the first method you might assert on methods that don't exist because the whole api is mocked:
```
client._datasets_api.some_other_method.assert_called_with(...)
```